lib/media_types/scheme.rb

```
@@ -158,11 +209,10 @@ def validate!(output, call_options, **_opts)
     #   MyMedia.valid?({ foo: { bar: 'my-string' }})
     #   # => true
     #
-    def attribute(key, type = ::Object, optional: false, **opts, &block)
-      raise KeyTypeError, "Unexpected key type #{key.class.name}, please use either a symbol or string." unless key.is_a?(String) || key.is_a?(Symbol)
-      raise DuplicateKeyError, "An attribute with key #{key.to_s} has already been defined. Please remove one of the two." if rules.has_key?(key)
-      raise DuplicateKeyError, "A string attribute with the same string representation as the symbol :#{key.to_s} already exists. Please remove one of the two." if key.is_a?(Symbol)&& rules.has_key?(key.to_s)
-      raise DuplicateKeyError, "A symbol attribute with the same string representation as the string '#{key}' already exists. Please remove one of the two." if key.is_a?(String) && rules.has_key?(key.to_sym)
+    def attribute(key, type = nil, optional: false, **opts, &block)
+      raise ConflictingTypeDefinitionError, 'You cannot apply a block to a typed attribute, either remove the type or the block' if block_given? && !type.nil?
+
+      type ||= ::Object
```

Why was this changed from the initializer in the method param?

---

The preceding if check on the raise can't check if type equals ::Object, because it is possible for the user to supply that as an actual argument. And we don't want it to break in that case

---

But type will never be nil in that case?

---

It should be when the user provides a block.

---

When an user provides a block to 'attribute', any given type is ignored. It was decided that doing so silently could cause confusion, so we want to raise an error (when the user gives both a block and a type).
If we set the default of type to 'Object', we must write an 'if' that checks if type != Object to detect if the user entered a type. However, 'Object' could be a valid type that the user could enter, so we changed the default to be nil (still something the user could enter, but would make much less sense to do so).
Now we're left with the fact that type is an optional parameter. So 'attribute :foo' should add the same rule as it did previously. With the default being now 'nil', that is no longer the case, so we should update the default to 'Object'

---

Got it. That's fine then.